src/rdb.c

```diff
@@ -2183,6 +2224,182 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, int dbid, int *error) {
+            * the hash fields there are, since the log of operations in the incr AOF
+            * is assumed to work in the exact keyspace state.
+            * Expired hash fields on the master are silently discarded.
+            * Note that if all fields in a hash has expired, the hash would not be
+            * created in memory (because it is created on the first valid field), and
+            * thus the key would be discarded as an "empty key" */
+            if (ttl != 0 && iAmMaster() && ((mstime_t)ttl < now)) { /* note: TTL was saved to RDB as unix-time in milliseconds */
+                /* TODO: consider replication (like in rdbLoadAddKeyToDb) */
+                server.rdb_last_load_hash_fields_expired++;
+                sdsfree(field);
+                sdsfree(value);
+                continue;
+            }
+
+            /* keep the nearest expiration to connect listpack object to db expiry */
+            if (ttl < minExpire) minExpire = ttl;
+
+            /* store the values read - either to listpack or dict */
+            if (o->encoding == OBJ_ENCODING_LISTPACK_EX) {
+                /* integrity - check for key duplication (if required) */
+                if (dupSearchDict) {
+                    sds field_dup = sdsnewlen(field, sdslen(field));
+
+                    if (dictAdd(dupSearchDict, field_dup, NULL) != DICT_OK) {
+                        rdbReportCorruptRDB("Hash with dup elements");
+                        dictRelease(dupSearchDict);
+                        if (hashAddCtx != NULL) hashTypeGroupSetDone(hashAddCtx);
+                        decrRefCount(o);
+                        sdsfree(field_dup);
+                        sdsfree(value);
+                        sdsfree(field);
+                        return NULL;
+                    }
+                }
+
+                /* check if the values can be saved to listpack (or should convert to dict encoding) */
+                if (sdslen(field) > server.hash_max_listpack_value ||
+                    sdslen(value) > server.hash_max_listpack_value ||
+                    lpEntrySizeInteger(ttl) > server.hash_max_listpack_value ||
+                    !lpSafeToAdd(((listpackEx*)o->ptr)->lp, sdslen(field) + sdslen(value) + lpEntrySizeInteger(ttl)))
+                {
+                    /* convert to hash */
+                    hashTypeConvert(db, o, OBJ_ENCODING_HT);
+                    hashAddCtx = HashTypeGroupSetInit(key, o, db);
+                    if (hashAddCtx == NULL) {
+                        decrRefCount(o);
+                        if (dupSearchDict != NULL) dictRelease(dupSearchDict);
+                        sdsfree(value);
+                        sdsfree(field);
+                        return NULL;
+                    }
+
+                    if (len > DICT_HT_INITIAL_SIZE) { /* TODO: this is NOT the original len, but this is also the case for simple hash, is this a bug? */
```

Comment:
My TODO is about len being decreased every iteration, so the expansion is done only for the reaming size, but this just might be the API definition (i.e., the `size` parameter is by how much to expand, not the total size).
I'll delete it.

Reply:
Indeed, `len` not truly represent number of tuples. But there is no bug, only inefficiency.  
Should be a distinction between `nTuples` and `nTuplesLeft`. 
Please consider either having it as part of this PR if you don't see any risk. 
Otherwise, let's have it on a distinct PR.